File: src/lib/scala/Slate.Integration/src/main/scala/slate/integration/EntitiesApi.scala
/**
<slate_header>
  url: www.slatekit.com
  git: www.github.com/code-helix/slatekit
  org: www.codehelix.co
  author: Kishore Reddy
  copyright: 2016 CodeHelix Solutions Inc.
  license: refer to website and/or github
  about: A Scala utility library, tool-kit and server backend.
  mantra: Simplicity above all else
</slate_header>
  */

package slate.integration

import slate.common.Result
import slate.common.databases.DbConString
import slate.core.apis.{Api, ApiAction, ApiBase}
import slate.entities.models.{EntitySetupService, ModelSettings}


@Api(area = "sys", name = "models", desc = "api to access and manage data models",
  roles= "admin", auth = "key-roles", verb = "post", protocol = "cli")
class EntitiesApi(val _settings:ModelSettings) extends ApiBase
{
  @ApiAction(name = "", desc = "installs the model to the database shard", roles= "@parent", verb = "@parent", protocol = "@parent")
  def install(name:String, version:String = "", dbKey:String = "", dbShard:String = "")
    : Result[Boolean] =
  {
    service.install(name, version, dbKey, dbShard)
  }


  @ApiAction(name = "", desc = "installs all the models in the default database", roles= "@parent", verb = "@parent", protocol = "@parent")
  def installAll(): Result[String] =
  {
    service.installAll()
  }


  @ApiAction(name = "", desc = "installs all the models in the default database", roles= "@parent", verb = "@parent", protocol = "@parent")
  def names(): List[String] =
  {
    service.names()
  }


  @ApiAction(name = "", desc = "generates sql install files for the model", roles= "@parent", verb = "@parent", protocol = "@parent")
  def generateSql(name:String, version:String = ""): Result[String] =
  {
    service.generateSql(name, version)
  }


  @ApiAction(name = "", desc = "generates sql install files for all models", roles= "@parent", verb = "@parent", protocol = "@parent")
  def generateSqlAll(): Result[String] =
  {
    service.generateSqlAll()
  }


  @ApiAction(name = "", desc = "gets the default db connection", roles= "@parent", verb = "@parent", protocol = "@parent")
  def connectionByDefault(): Result[DbConString] =
  {
    service.connectionByDefault()
  }


  @ApiAction(name = "", desc = "gets the default db connection", roles= "@parent", verb = "@parent", protocol = "@parent")
  def connectionByName(name:String): Result[DbConString] =
  {
    service.connectionByName(name)
  }


  @ApiAction(name = "", desc = "get the database connection for the db shard", roles= "@parent", verb = "@parent", protocol = "@parent")
  def connectionByGroup(dbKey:String = "", dbShard:String = ""): Result[DbConString] =
  {
    service.connectionByGroup(dbKey, dbShard)
  }


  private def service: EntitySetupService =
  {
    new EntitySetupService(context.ent, context.dbs, _settings, context.dirs)
  }
}

File: src/support/integration/src/main/kotlin/kiit/integration/apis/EntitiesApi.kt
/**
 * <slate_header>
 * url: www.slatekit.com
 * git: www.github.com/code-helix/slatekit
 * org: www.codehelix.co
 * author: Kishore Reddy
 * copyright: 2016 CodeHelix Solutions Inc.
 * license: refer to website and/or github
 * about: A tool-kit, utility library and server-backend
 * mantra: Simplicity above all else
 * </slate_header>
 */

package kiit.integration.apis

import kiit.apis.Api
import kiit.apis.Action
import kiit.apis.AuthModes
import kiit.apis.Verbs
import kiit.apis.ApiBase
import kiit.common.Sources
import kiit.common.data.DbCon
import kiit.common.newline
import kiit.common.data.Connections
import kiit.connectors.entities.AppEntContext
import kiit.migrations.MigrationService
import kiit.migrations.MigrationSettings
import kiit.results.Notice
import kiit.results.Try

@Api(area = "infra", name = "entities", desc = "api to access and manage data models",
        auth = AuthModes.KEYED, roles = ["admin"], verb = Verbs.AUTO, sources = [Sources.ALL])
class EntitiesApi(context: AppEntContext) : ApiBase(context) {
    val appContext = context

    @Action(desc = "installs the model to the database shard")
    fun install(name: String, version: String = "", dbKey: String = "", dbShard: String = ""): Try<String> {
        return service().install(name, version, dbKey, dbShard)
    }

    @Action(desc = "installs all the models in the default database")
    fun installAll(): Try<List<String>> {
        return service().installAll()
    }

    @Action(desc = "deletes data in the table supplied")
    fun delete(name: String): Try<String> {
        return service().delete(name)
    }

    @Action(desc = "deletes data in all the tables")
    fun deleteAll(): Try<List<String>> {
        return service().deleteAll()
    }

    @Action(desc = "drops the table supplied")
    fun drop(name: String): Try<String> {
        return service().drop(name)
    }

    @Action(desc = "dropss the table")
    fun dropAll(): Try<List<String>> {
        return service().dropAll()
    }

    @Action(desc = "installs all the models in the default database")
    fun names(): List<Pair<String, String>> {
        val service = service()
        return service.names()
    }

    @Action(desc = "installs all the models in the default database")
    fun counts(): List<Pair<String, Long>> {
        return service().counts()
    }

    @Action(desc = "generates sql install files for the model")
    fun generateSql(name: String, version: String = ""): Try<String> {
        return service().generateSql(name, version).map { it.joinToString(newline) }
    }

    @Action(desc = "generates sql install files for all models")
    fun generateSqlFiles(): Try<List<String>> {
        return service().generateSqlFiles()
    }

    @Action(desc = "generates a single sql install file for all models")
    fun generateSqlAllInstall(): Try<String> {
        return service().generateSqlAllInstall()
    }

    @Action(desc = "generates a single sql install file for all models")
    fun generateSqlAllUninstall(): Try<String> {
        return service().generateSqlAllUninstall()
    }

    @Action(desc = "gets the default db connection")
    fun connection(): Notice<DbCon> {
        return service().connection()
    }

    @Action(desc = "gets the default db connection")
    fun connectionByName(name: String): Notice<DbCon> {
        return service().connectionByName(name)
    }

    private fun service(): MigrationService {
        return MigrationService(appContext.ent, appContext.dbs, MigrationSettings(), context.dirs)
    }
}
